x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java

```diff
@@ -125,24 +125,24 @@ else if (p.resolved()) {
                 agg.aggregates().forEach(e -> {
                     var exp = e instanceof Alias ? ((Alias) e).child() : e;
                     if (exp instanceof AggregateFunction aggFunc) {
-                        aggFunc.arguments().forEach(a -> {
-                            // TODO: allow an expression?
-                            if ((a instanceof FieldAttribute
-                                || a instanceof MetadataAttribute
-                                || a instanceof ReferenceAttribute
-                                || a instanceof Literal) == false) {
-                                failures.add(
-                                    fail(
-                                        e,
-                                        "aggregate function's parameters must be an attribute or literal; found ["
-                                            + a.sourceText()
-                                            + "] of type ["
-                                            + a.nodeName()
-                                            + "]"
-                                    )
-                                );
-                            }
-                        });
+                        Expression field = aggFunc.field();
```

Comment:
This leaves the non-field argument verification to each AggregateFunction in part. There are so far just two of these that require an extra param: `CountDistinct` and `Percentile`. I think there's no language restriction that these extra parameters cannot be a some sort of a field and this PR removes that check. However, both of these currently do expect a literal. So I guess relaxing this verifier check is OK.
However, there's a bug in `CountDistinct`, its type resolution not checking if the 2nd parameter is foldable, but then trying to fold it. This check was guarding against that, but with this change we should also fix that -- up to you if you want to do it here or subsequent PR.

Reply:
That's a good observation. I'll make sure the type resolution in `CountDistinct` is fixed and checked by tests.